### DISCUSSION TITLE
How to conditionally require a field?

### DISCUSSION THREAD

**Discussion by ewautr** [open-circle/valibot#297] [Q&A] [answered]
Hi, I have a field that should be required conditionally, depending on a checkbox outside of the form. How can I implement that with Valibot?

`const schema = object(
  {
    name: nullable(string([minLength(1)])),
  },
);`

**Comment by fabian-hiller** [comment]
There are several solutions. To make it type safe, I recommend using a discriminated union with `variant`. To make it easy, I recommend using `custom` in the pipeline of your `object` to validate the data type or value of the key. If the dependent value is not part of the schema, you will have to figure out how to get it into the schema validation. If you provide me with more code, I can try to help you.

**Reply by ewautr** [reply]
Thanks for answering, it would be awesome to see an example of a solution. Here is some code with my schema and a reactive variable I need to use. If a variable is true, the name field should be a required string, if it is false it should be null. 

```
const nameUnknown = ref(false);

const schema = object({
  name: nullable(string([minLength(1)])),
});
```

**Reply by fabian-hiller** [reply, marked as answer]
I recommend to write the following. `union` has no `pipe` argument yet, that's why I added the validation to `object` and forwarded the error to `name`. I plan to improve that.

> Whenever `nameUnknown.value` is `true` or `name` is a string, the input is valid. Otherwise the error message is displayed.

```ts
import { custom, forward, minLength, null_, nullable, object, string, union } from 'valibot';

const nameUnknown = ref(false);

// Currently
const schema = object(
  {
    name: nullable(string([minLength(1)])),
  },
  [
    forward(
      custom(
        (input) => nameUnknown.value || typeof input.name === 'string',
        'Name is required.'
      ),
      ['name']
    ),
  ]
);

// Planned improvement
const schema = object({
  name: union(
    [null_(), string([minLength(1)])],
    [
      custom(
        (input) => nameUnknown.value || typeof input.name === 'string',
        'Name is required.'
      ),
    ]
  ),
});
```

**Reply by ewautr** [reply]
Works great, thank you for help!

**Reply by fabian-hiller** [reply]
Please let me know if you have any ideas on how to improve the API and DX of Valibot. I am happy to receive any feedback!

**Reply by fabian-hiller** [reply]
The planned improvement will be available in the next release.

**Reply by fabian-hiller** [reply]
v0.25.0 is available

**Reply by ewautr** [reply]
Btw one more question about the solution you suggested. I get some errors when trying to add some typing to the input argument. Do you know why that could be?
Two main error I am getting is '_No overload matches this call._' and '_Type '"flightHours"' is not assignable to type '"message" | "requirement" | "async" | "_parse" | "type"'_'

Here is my schema:
`const schema = object(
  {
    flightHours: nullish(string()),
  },
  [
    forward(
      custom(
        (input: CustomValidation<FormType>) => isFlyingHoursNotRequired.value ||
          (typeof input.flightHours === 'string' && input.flightHours.length > 0),
        'Flying hours are required.',
      ),
      ['flightHours'],
    )
  ],
);`

Form type:
`export interface RequestExchangeForm {
  flightHours?: string;
}
`

**Reply by fabian-hiller** [reply]
Why do you want to add your own types to the input argument of the function in `custom`? It is already fully typed by my Valibot.

**Reply by fresonn** [reply]
is it possible to implement that with `variant`?
Are there any examples in the documentation?

**Reply by fabian-hiller** [reply]
If the `variant` depends on an external value, probably not. `lazy` might work in this case. Scroll down on the following pages for examples.
- https://valibot.dev/api/variant/
- https://valibot.dev/api/lazy/

**Comment by fresonn** [comment]
@fabian-hiller I'll continue here ok?


This is how I solved this problem, variant and union didn't help

```typescript
import * as v from 'valibot';

export const schema1 = v.object({
  name: v.pipe(
    v.string(),
    v.trim(),
    v.nonEmpty(),
  ),
});

/*

// This won't work

const schema2 = v.object({
  supplier: v.picklist(['JS', 'GO']),
  ecmaVersion: v.union(
    [v.null(), v.string()],
    [v.check((input) = input.supplier !== 'JS', 'ecmaVersion required'],
  ),
});
*/

export const schema2 = v.pipe(
  v.object({
    supplier: v.picklist(['JS', 'GO']),
    ecmaVersion: v.optional(v.string())
  }),
  v.forward(
    v.check((input) => input.supplier !== 'JS', 'ecmaVersion required'),
    ['ecmaVersion']
  )
)

// This won't work
// const Schema = v.object({
//   ...schema1.entries,
//   ...schema2.entries,
// });

const Schema = v.intersect([
  schema1,
  schema2,
]);



const result = v.safeParse(Schema, {
  name: 'Some name',
  supplier: 'JS',
  recurrence: undefined,
});

console.log(result);
```

**Reply by fabian-hiller** [reply]
Yes, our API has changed since this discussion was created: https://valibot.dev/blog/valibot-v0.31.0-is-finally-available/

Thank you for sharing your working code with others here!

**Reply by fresonn** [reply]
@fabian-hiller thank you for your library too
